starlark/src/eval/mod.rs

```diff
@@ -203,16 +203,59 @@ pub trait FileLoader: 'static {
     fn load(&self, path: &str) -> Result<Environment, EvalException>;
 }
 
+/// Starlark `def` function local variables
+pub(crate) struct DefLocals<'a> {
+    // name to index is needed for nested contexts only
+    // (collection comprehensions).
+    // TODO: access nested context objects by slot index too and drop this field
+    name_to_index: &'a HashMap<String, usize>,
+    locals: RefCell<Vec<Option<Value>>>,
```

Comment:
I would add an extensive documentation to this variable explaining the schema, something like

```rust
/// locals store the local variable in an array. Each local variable loose its names and the `name_to_index`
/// can be used to find a variable from name. Once the parsing of the `def` is finished, only the nested contexts should
/// need it (the other one have been replaced during the parsing).
/// `locals` is used to optimize access speed to local variable (versus looking into a hashmap).
```

Also maybe add some more explanation to the `transform_locals` method? I am a bit worried that this is a tricky optimization and we will forget what it is for.

Reply:
That's well-written explanation. Thank you!